### PR TITLE
[Curp] Refactor sync manager and add shutdown channel

### PR DESCRIPTION
**Pull request by markcty**
Refactor sync manager and add shutdown broadcast channel.

### CODE REVIEW

File: curp/src/channel/mod.rs [hunk 1]
```diff
@@ -180,15 +180,13 @@ impl<KM> KeyBasedReceiverInner<KM> {
         let mut listener = None;
         loop {
             channel.flush_queue(&mut self.buf);
-            match self.buf.pop_front() {
```
Comment (markcty) [inline curp/src/channel/mod.rs:183]: The clippy in CI has problem with these lines. But my local clippy keeps warning me `single_match_else` here. So I refactor these lines. I guess we need to update the clippy version in our CI

File: curp/src/server.rs [hunk 2]
```diff
@@ -734,8 +745,8 @@ impl<C: 'static + Command, CE: 'static + CommandExecutor<C>> Protocol<C, CE> {
 
         loop {
             let fr_or_listener = self.cmd_board.map_lock(|mut board| {
-                let fr_or_listener = match board.get(&id) {
```
Comment (markcty) [inline curp/src/server.rs:737]: Same clippy `single_match_else` problem as above

File: curp/src/channel/mod.rs (hunk 1 above)
Comment (rogercloud) [reply curp/src/channel/mod.rs:183]: You can try to check the clippy version locally.

File: curp/src/server.rs [hunk 3]
```diff
@@ -274,15 +273,11 @@ pub struct Protocol<C: Command + 'static, CE: CommandExecutor<C> + 'static> {
...
     sync_chan: MpscKeyBasedSender<C::K, SyncMessage<C>>,
-    /// Synced manager handler
-    sm_handle: JoinHandle<()>,
-    /// After sync tasks handle
-    after_sync_handle: Vec<JoinHandle<()>>,
     // TODO: clean up the board when the size is too large
     /// Cmd watch board for tracking the cmd sync results
     cmd_board: Arc<Mutex<HashMap<ProposeId, CmdBoardValue>>>,
-    /// trigger when commit_index changes
-    commit_trigger: UnboundedSender<()>,
```
Comment (markcty) [inline curp/src/server.rs:285]: The commit_trigger is also moved into state to keep consistent with other triggers.

File: curp/src/bg_tasks.rs [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+                    }
+                } else {
+                    // followers execute commands directly
+                    let cmd_executor = Arc::clone(&cmd_executor);
+                    let cmd = Arc::clone(cmd);
+                    let spec = Arc::clone(&spec);
+                    let _handle = tokio::spawn(async move {
+                        // TODO: execute command in parallel
+                        // TODO: handle execution error
+                        let _execute_result = cmd.execute(cmd_executor.as_ref()).await;
```
Comment (rogercloud) [inline curp/src/bg_tasks.rs:406]: Can we make sure the first-come command executed first? What if there are conflicted commands?

File: curp/src/channel/mod.rs (hunk 1 above)
Comment (rogercloud) [reply curp/src/channel/mod.rs:183]: Can we fix the clippy issue?

File: curp/src/shutdown.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    pub(crate) fn new(notify: broadcast::Receiver<()>) -> Shutdown {
+        Shutdown {
+            shutdown: false,
+            notify,
+        }
+    }
+
+    /// Returns `true` if the shutdown signal has been received.
+    #[allow(dead_code)]
+    pub(crate) fn is_shutdown(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (markcty) [reply curp/src/shutdown.rs:23]: Removed

File: curp/src/bg_tasks.rs [hunk 6]
```diff
@@ -0,0 +1,635 @@
...
+        self.log_index
+    }
+
+    /// Get commands
+    pub(crate) fn cmd(&self) -> Arc<C> {
+        Arc::clone(&self.cmd)
+    }
+}
+
+/// The message sent to the `SyncManager`
```
[the remark this answers is not in the document]
Comment (markcty) [reply curp/src/bg_tasks.rs:98]: Corrected

File: curp/src/bg_tasks.rs [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+/// Send `append_entries` containing a single log to a server
+#[allow(clippy::integer_arithmetic, clippy::indexing_slicing)] // log.len() >= 1 because we have a fake log[0], indexing of `next_index` or `match_index` won't panic because we created an entry when initializing the server state
+async fn send_log_until_succeed<C: Command + 'static>(
+    i: usize,
+    connect: Arc<Connect>,
+    state: Arc<RwLock<State<C>>>,
+) {
+    // prepare append_entries request args
+    #[allow(clippy::shadow_unrelated)] // clippy false positive
+    let (term, prev_log_index, prev_log_term, entries, leader_commit) = state.map_read(|state| {
```
[the remark this answers is not in the document]
Comment (markcty) [reply curp/src/bg_tasks.rs:222]: Fixed

File: curp/src/bg_tasks.rs [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+
+/// Fetch commands need to be synced and add them to the log
+async fn bg_get_sync_cmds<C: Command + 'static>(
+    state: Arc<RwLock<State<C>>>,
+    mut sync_chan: MpscKeyBasedReceiver<<C as Command>::K, SyncMessage<C>>,
+    ae_trigger: UnboundedSender<usize>,
+) {
+    loop {
+        let (cmds, term) = match fetch_sync_msgs(&mut sync_chan).await {
+            Ok((cmds, term)) => (cmds, term),
```
[the remark this answers is not in the document]
Comment (markcty) [reply curp/src/bg_tasks.rs:132]: Created.

File: curp/src/bg_tasks.rs (hunk 4 above)
Comment (markcty) [reply curp/src/bg_tasks.rs]: Fixed in new commit.

File: curp/src/channel/mod.rs (hunk 1 above)
Comment (markcty) [reply curp/src/channel/mod.rs:183]: Fixed.

File: curp/src/cmd_board.rs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        Self {
+            notifiers: HashMap::new(),
+            cmd_states: HashMap::new(),
+        }
+    }
+}
+
+/// The state of a command in cmd watch board
+/// (`EarlyArrive` -> ) `Execute` -> `AfterSync` -> `FinalResponse`
+// TODO: this struct might me removed. We don't need to store whether the command needs execution after sync in one place. We can attach it to SyncMessage.
```
[the remark this answers is not in the document]
Comment (markcty) [reply curp/src/cmd_board.rs:28]: Fixed.

File: curp/src/bg_tasks.rs [hunk 10]
```diff
@@ -0,0 +1,694 @@
...
+        }
+    }
+}
+
+/// The only place where cmds get executed and call `after_sync`
+async fn bg_execute_cmd<C: Command + 'static, CE: 'static + CommandExecutor<C>>(
+    ce: CE,
+    mut cmd_rx: mpsc::UnboundedReceiver<ExecuteMessage<C>>,
+) {
+    // TODO: use KeyBasedMpsc to dispatch cmds to execute in parallel
```
[the remark this answers is not in the document]
Comment (markcty) [reply curp/src/bg_tasks.rs:675]: Added.

File: curp/src/cmd_execute_worker.rs [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+                    Some(cmd.after_sync(ce.as_ref(), index).await)
+                } else {
+                    None
+                };
+                let _ignore = tx.send((er, asr)); // it's ok to ignore the result here because sometimes the result is not needed
+            }
+        }
+
+        if let Err(e) = done.send(msg_wrapped) {
+            warn!("{e}");
```
[the remark this answers is not in the document]
Comment (markcty) [reply curp/src/cmd_execute_worker.rs:52]: Fixed.

**Review by rogercloud — APPROVED** [review]

**Pull request merged**